src/adapter.rs

```diff
@@ -0,0 +1,739 @@
+    }
+    fn get_discoverable_timeout(&self)-> Result<u32, Box<Error>>;
+    fn set_discoverable_timeout(&self, timeout: u32) -> Result<(), Box<Error>>{
+        Err(Box::from(NOT_SUPPORTED_ERROR))  
+    } 
+    fn is_discovering(&self)-> Result<bool, Box<Error>>;
+    fn set_discovering(&self, discovering: bool) -> Result<(), Box<Error>>{
+        Err(Box::from(NOT_SUPPORTED_ERROR))  
+    } 
+    fn set_can_start_discovery(&self, can_start_discovery: bool) -> Result<(), Box<Error>>{
+        Err(Box::from(NOT_SUPPORTED_ERROR))  
+    }
+    fn set_can_stop_discovery(&self, can_stop_discovery: bool) -> Result<(), Box<Error>>{
+        Err(Box::from(NOT_SUPPORTED_ERROR))  
+    }
+    fn create_discovery_session(&self)-> Result<BluetoothDiscoverySession, Box<Error>> ;
+    fn get_uuids(&self)-> Result<Vec<String>, Box<Error>>;
+    fn set_uuids(&self, uuids: Vec<String>) -> Result<(), Box<Error>>{
+        Err(Box::from(NOT_SUPPORTED_ERROR))  
+    } 
+    fn get_vendor_id_source(&self)-> Result<String, Box<Error>>;
+    fn get_vendor_id(&self)-> Result<u32, Box<Error>>;
+    fn get_product_id(&self) -> Result<u32, Box<Error>> ;
+    fn get_device_id(&self) -> Result<u32, Box<Error>>;
+    fn get_modalias(&self) -> Result<(String, u32, u32, u32), Box<Error>>;  
+    fn set_modalias(&self, modalias: String) -> Result<(), Box<Error>>{
+        Err(Box::from(NOT_SUPPORTED_ERROR))  
+    }
+    fn get_ad_datas(&self) -> Result<Vec<String>, Box<Error>>{
+        Err(Box::from(NOT_SUPPORTED_ERROR))  
+    }
+    fn set_ad_datas(&self, ad_datas: Vec<String>) -> Result<(), Box<Error>>{
+        Err(Box::from(NOT_SUPPORTED_ERROR))  
+    }
+}
+
+impl BluetoothAdapter{
+    #[cfg(all(target_os = "linux", feature = "bluetooth"))]
+    pub fn new() -> Result<Box<BluetoothAdapterBluez>, Box<Error>> {
+        let bluez_adapter = try!(BluetoothAdapterBluez::init());
+        Ok(Box::new(bluez_adapter))
+    }
+
+    #[cfg(all(target_os = "android", feature = "bluetooth"))]
+    pub fn new() -> Result<Box<BluetoothAdapterAndroid>, Box<Error>> {
+        let blurdroid_adapter = try!(BluetoothAdapterAndroid::get_adapter());
+        Ok(Box::new(blurdroid_adapter))
+    }
+
+    #[cfg(all(target_os = "macos", feature = "bluetooth"))]
+    pub fn new() -> Result<Box<BluetoothAdapterMac>, Box<Error>> {
+        let mac_adapter = try!(BluetoothAdapterMac::init());
+        Ok(Box::new(mac_adapter))
+    }
+
+    #[cfg(not(any(all(target_os = "linux", feature = "bluetooth"),
+                  all(target_os = "android", feature = "bluetooth"),
+                  all(target_os = "macos", feature = "bluetooth"))))]
+    pub fn new() -> Result<Box<BluetoothAdapterEmpty>, Box<Error>> {
+        let adapter = try!(BluetoothAdapterEmpty::init());
+        Ok(Box::new(adapter))
+    }
+
+    #[cfg(feature = "bluetooth-test")]
+    pub fn new_mock() -> Result<Box<FakeBluetoothAdapter>, Box<Error>> {
+        Ok(Box::new_empty())
```

Comment:
This line isn't right. It should be calling Box::new(...) with an instance of FakeBluetoothAdapter, and the return type should be `Box<BluetoothAdapter>`.

Reply:
When I tried to fix this by making the changes you suggested, it says trait bluetoothadapter is not implemented for blurmac : : FakebluetoothAdapter (But I cannot make changes to that file right?)

The changes we made for this method in adapter.rs is,
pub fn new_mock() -> Result<Box<BluetoothAdapter, Box<Error>> {
        let fake = FakeBluetoothAdapter::new_empty();
        Ok(Box::new(fake))
    }

Reply:
Or are we approaching it the wrong way?

Reply:
Oh, I see what's happening. Like the `struct Mac` and `struct Android`, you need a struct that wraps the FakeBluetoothAdapter so you can implement the BluetoothAdapter trait for it, and return that.

Reply:
My mistake; there already is a `struct Mock`. That's what you should be returning inside of the Box.

Reply:
Okay so if I do that, then I have to pass an instance of Mock on Box::new(...) right? 
Because it now expects struct 'adapter : : Mock'

Reply:
Yes.